src/main/drivers/pwm_output_dshot_hal.c

```diff
@@ -58,233 +58,244 @@ void pwmWriteDshotInt(uint8_t index, uint16_t value)
-        /* Set the DMAR specific dma handle parameters to be configured */
-        motor->hdma_tim.Init.Channel = timerHardware->dmaTimUPChannel;
-        motor->hdma_tim.Init.FIFOMode = DMA_FIFOMODE_ENABLE;
-
-        /* Set hdma_tim instance */
-        motor->hdma_tim.Instance = timerHardware->dmaTimUPRef;
+        motor->timer->dmaBurstRef = dmaRef;
 
-        /* Link hdma_tim to hdma[x] (channelx) */
-        __HAL_LINKDMA(&motor->TimHandle, hdma[motor->timerDmaIndex], motor->hdma_tim);
-
-        dmaInit(timerHardware->dmaTimUPIrqHandler, OWNER_TIMUP, timerGetTIMNumber(timerHardware->tim));
-        dmaSetHandler(timerHardware->dmaTimUPIrqHandler, motor_DMA_IRQHandler, NVIC_BUILD_PRIORITY(1, 2), motorIndex);
+        if (!configureTimer) {
+            motor->configured = true;
+            return;
+        }
     } else
 #endif
     {
-        motor->timerDmaIndex = timerDmaIndex(timerHardware->channel);
-        motor->timer->timerDmaSources |= timerDmaSource(timerHardware->channel);
-
-        /* Set the non-DMAR specific dma handle parameters to be configured */
-        motor->hdma_tim.Init.Channel = timerHardware->dmaChannel;
-        motor->hdma_tim.Init.FIFOMode = DMA_FIFOMODE_DISABLE;
-
-        /* Set hdma_tim instance */
-        motor->hdma_tim.Instance = timerHardware->dmaRef;
-
-        /* Link hdma_tim to hdma[x] (channelx) */
-        __HAL_LINKDMA(&motor->TimHandle, hdma[motor->timerDmaIndex], motor->hdma_tim);
-
-        dmaInit(timerHardware->dmaIrqHandler, OWNER_MOTOR, RESOURCE_INDEX(motorIndex));
-        dmaSetHandler(timerHardware->dmaIrqHandler, motor_DMA_IRQHandler, NVIC_BUILD_PRIORITY(1, 2), motorIndex);
-    }
-
-    /* Initialize TIMx DMA handle */
-    if (HAL_DMA_Init(motor->TimHandle.hdma[motor->timerDmaIndex]) != HAL_OK) {
-        /* Initialization Error */
-        return;
+        motor->timerDmaSource = timerDmaSource(timerHardware->channel);
+        motor->timer->timerDmaSources &= ~motor->timerDmaSource;
     }
 
-    TIM_OC_InitTypeDef TIM_OCInitStructure;
-
-    /* PWM1 Mode configuration: Channel1 */
-    TIM_OCInitStructure.OCMode = TIM_OCMODE_PWM1;
-    TIM_OCInitStructure.OCIdleState = TIM_OCIDLESTATE_SET;
-    TIM_OCInitStructure.OCPolarity = (output & TIMER_OUTPUT_INVERTED) ? TIM_OCPOLARITY_LOW : TIM_OCPOLARITY_HIGH;
-    TIM_OCInitStructure.OCNIdleState = TIM_OCNIDLESTATE_SET;
-    TIM_OCInitStructure.OCNPolarity = (output & TIMER_OUTPUT_INVERTED) ? TIM_OCNPOLARITY_LOW : TIM_OCNPOLARITY_HIGH;
-    TIM_OCInitStructure.OCFastMode = TIM_OCFAST_DISABLE;
-    TIM_OCInitStructure.Pulse = 0;
-
-    if (HAL_TIM_PWM_ConfigChannel(&motor->TimHandle, &TIM_OCInitStructure, motor->timerHardware->channel) != HAL_OK) {
-        /* Configuration Error */
-        return;
-    }
+    DMA_TypeDef *dma;
+    uint32_t stream;
+    if (dmaRef == DMA1_Stream0) { dma = DMA1; stream = LL_DMA_STREAM_0; }
+    else if (dmaRef == DMA1_Stream1) { dma = DMA1; stream = LL_DMA_STREAM_1; }
```

Comment:
Should this be a `switch`?

Reply:
Switch on addresses feels weird, but it probably could.

It's one of the things gluing LL and our code together, I feel like I either should rewrite LL functions to allow both DMA+Stream and DMA_Stream forms, to avoid possible overhead of runtime address arithmetic, or alter timer definitions to store both forms of DMA representation.

Reply:
Actually casting pointer back to label is quite hard .. you will probably end up using `DMA1_Stream1_BASE`

Also it should be possible to abstract this to DMA handling code .. Configurable DMA PR should contain most of the plumbing ...

Reply:
That's what I don't want to do, we could alter `TIMER_DEF` to store both `DMA_Stream_TypeDef*` and `DMA_TypeDef*`, `stream`, `channel` separately since all of our `D(d, s, c)` definitions in `timer_def.h` already have that info.

However I like leaving `DMA_Stream_TypeDef*` and rewriting LL-functions where necessary.
Configurable DMA is being worked on? I could have a look at it.